### ISSUE TITLE
Bug report: Closing tabs don't free up memory

### ISSUE THREAD

**Issue by yaichenbaum** [files-community/Files#2843]
**To Reproduce**
1. Open a couple tabs
2. Close them

**Expected behavior**
Memory goes down

**Comment by winston-de** [comment]
It looks like there are some circular refences to the ModernShellPage that may be preventing proper garbage collection.
Files.ViewModels.ItemViewModel and Files.BaseLayout both contain circular references to ModernShellPage.

**Comment by d2dyno1** [comment]
I feel like a lot of core logic has to be rewritten

**Comment by ziv-commentor** [comment]
I don't feel like this is a problem: The file explorer merely takes up 100 MB of RAM, opening up 20 tabs doesn't increase RAM usage either.

**Comment by yaichenbaum** [comment]
@ziv-commentor The tab itself takes around 1mb but if you navigate to large directory such as System 32, you will see the memory usage increase and closing the tab doesn't free up the memory.

**Comment by winston-de** [comment]
Related: 
https://github.com/microsoft/microsoft-ui-xaml/issues/3597
#2065

**Comment by duke7553** [comment]
@d2dyno1 Merely saying it needs to be re-written without specific solutions won't help anything. 

@winston-de Re-working ItemViewModel is quite a task, and I avoided it with the overhaul I did several months ago. We should aim to remove unwanted references from inside ItemViewModel. 

The minimal sample project I created on that WinUI issue demonstrates the issue is not entirely related to Files logic. I'd be interested in what is using the most memory after tabs are closed. TabView control uses ListView under the hood, yet clearing the data source doesn't free old memory (item containers?) either.

Old issue I found for reference: https://social.msdn.microsoft.com/Forums/en-US/92f76519-f139-4c94-ae5e-630f0abda66e/uwpcmemory-leak-with-listview-ui-virtualization

**Comment by winston-de** [comment]
> I'd be interested in what is using the most memory after tabs are closed. TabView control uses ListView under the hood, yet clearing the data source doesn't free old memory (item containers?) either.

@duke7553 

These classes were the top 5 biggest users of memory after opening 7 tabs of System32 and closing it again.

Object Type | Count | Count Diff. | Size (Bytes) | Total Size Diff. (Bytes) | Inclusive Size (Bytes) | Inclusive Size Diff. (Bytes) | Module
-- | -- | -- | -- | -- | -- | -- | --
Files.Filesystem.ListedItem | 17,333 | 17,333 | 8,920,852 | 8,920,852 | 9,504,672 | 9,504,672 | Files.exe
Container<Object,   EventRegistrationTokenListWithCount> | 18,506 | 18,217 | 3,405,616 | 3,352,440 | 3,405,616 | 3,352,408 | System.Private.CoreLib.dll
Microsoft.Toolkit.Uwp.UI.Controls.DataGridCell   [RCW] | 1,754 | 1,754 | 2,496,700 | 2,496,700 | 2,496,700 | 2,496,700 | Microsoft.Toolkit.Uwp.UI.Controls.DataGrid.dll
Microsoft.Toolkit.Uwp.UI.Controls.DataGridRow   [RCW] | 231 | 231 | 952,940 | 952,940 | 3,485,676 | 3,485,676 | Microsoft.Toolkit.Uwp.UI.Controls.DataGrid.dll
Files.Views.ModernShellPage [RCW] | 8 | 7 | 926,440 | 888,144 | 1,490,856 | 1,418,216 | Files.exe 


[Here's the full diagnostic report.](https://github.com/files-community/Files/files/5772854/Report20210105-1358.zip)
Snapshot one was taken at startup, snapshot two was taken after open 7 tabs of System32, and snapshot three was taken after closing those tabs.

**Comment by d2dyno1** [comment]
From the spreadsheet, it looks like, we need `IDisposable` for `ListedItem` - when navigating to a directory: dispose all items in `foreach` loop, clear the collection, navigate to directory

**Comment by winston-de** [comment]
I don't think it's necessary to make ListedItem implement IDisposable. 
Instead, we can clear the lists and observable collections and whatnot in the ```Dispose``` method for ```ItemViewModel```. That should clean up a fair amount of memory.

I also think clearing the data sources that the DataGrid uses should also clear up the rows and columns, but I'm not certain.

**Comment by lukeblevins** [comment]
> I also think clearing the data sources that the DataGrid uses should also clear up the rows and columns, but I'm not certain.

@yair100 @winston-de I think this is what #2900 did, but I noticed a slight difference only if any. Maybe our problem is that we register events for the Rows and GridViewItems, but never unregister them when they need to be unloaded. I'm curious whether you think that's an issue or not.

**Comment by winston-de** [comment]
@duke7553 I agree that event handlers and property changed notifications in the layouts are taking up significant amounts of memory, and are an issue.
I'd be curious to know how much of this is caused by the TabView memory leak.

**Comment by duke7553** [comment]
@winston-de The question is how should we best unsubscribe from the events registered for each GridViewItem and DataGridRow? I'm going to look into this today.

**Comment by d2dyno1** [comment]
@duke7553 you might be interested in that:

https://docs.microsoft.com/en-us/dotnet/api/system.weakreference?view=net-5.0

**Comment by winston-de** [comment]
Memory usage seems to go down now, it does take a few seconds after closing the tab though.

Is this the case for anyone else?

**Comment by yaichenbaum** [comment]
@winston-de Perhaps a little, but not to where it was before the tab was opened.

**Comment by yair100** [comment]
I'm pleased to share that this issue is resolved in Files Preview. https://files.community/appinstallers/Files.preview.appinstaller
